Decouple routing and primary operation logic in TransportReplicationAction

areek opened the pull request:
Currently the logic to route an operation to the primary and the local primary operation is tightly coupled in `TransportReplicationAction#PrimaryPhase`. This change decouples the routing and retrying logic to `ReroutePhase` and the primary operation to `PrimaryPhase`, allowing to perform more informed shard level operations in the future.

brwe commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -566,13 +534,78 @@ void finishOnRemoteSuccess(Response response) {
+                return;
+            }
+            if (observer.observedState().nodes().nodeExists(primary.currentNodeId()) == false) {
+                logger.trace("primary shard [{}] is assigned to anode we do not know the node, scheduling a retry.", primary.shardId(), primary.currentNodeId());
```
maybe 'primary shard is assigned to an unknown node...' ?

bleskes replied:
I don't think we need to check for existence, just if it is us, which is covered by the next line. Makes sense?

areek replied:
yes, thanks for the pointer, the check was redundant. fixed

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -427,76 +434,51 @@ protected boolean checkBlocks() {
+                    try {
+                        // if we got disconnected from the node, or the node / shard is not in the right state (being closed)
+                        if (exp.unwrapCause() instanceof ConnectTransportException || exp.unwrapCause() instanceof NodeClosedException ||
+                                retryPrimaryException(exp.unwrapCause())) {
```
If we're doing a reroute - I don't think we should retry on retryPrimaryException. That one only holds for the primary action.

bleskes replied:
to clarify , the retryPrimaryException is only relevant for the handler of the primary action, not the reroute one. If it comes back from a reroute it means the target node already retried it and we can just fail.

areek replied:
Currently, retryPrimaryException returns true if the exception was caused by delay in dynamic mappings propagation (`RetryOnPrimaryException`) to the primary or when shard is unavailable (`TransportActions.isShardNotAvailableException`). 
I understand why we should not retry the reroute phase when we get `RetryOnPrimaryException` but IMO we should still retry the reroute phase when we see any `TransportActions.isShardNotAvailableException` exceptions. This is thrown by the reroute phase when the primary is not yet active or has not made it to the routing table. Thoughts?

areek replied:
agreed. the reroute phase will retry itself on hitting any unavailable shard exceptions and RetryOnPrimary and unavailable shard exceptions should only be retried coming from the `transportPrimaryAction`

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -427,76 +434,51 @@ protected boolean checkBlocks() {
-                }
-            }
-            return null;
+        protected void performReroute(final ShardRouting primary) {
```
do we really need these one liner methods? can we just inline what they do?

areek replied:
Added them to make them more readable w.r.t. what they do, I reverted it and added comments instead

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -581,7 +614,7 @@ void performOnPrimary(final ShardRouting primary, final ShardIterator shardsIt)
+                replicationPhase = new ReplicationPhase(shardsIt, primaryResponse.v2(), primaryResponse.v1(), observer, primary, internalRequest, channel, indexShardReference, shardFailedTimeout);
             } catch (Throwable e) {
                 // shard has not been allocated yet, retry it here
                 if (retryPrimaryException(e)) {
```
doesn't seem like we need this clause any more, it will be re-tried in the routing phase, right?

areek replied:
yes also because re-trying the routing phase will create a new reference to `indexShardReference`, will remove.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -427,76 +436,43 @@ protected boolean checkBlocks() {
+                    try {
+                        // if we got disconnected from the node, or the node / shard is not in the right state (being closed)
+                        if (exp.unwrapCause() instanceof ConnectTransportException || exp.unwrapCause() instanceof NodeClosedException ||
+                                (action.equals(transportPrimaryAction) && retryPrimaryException(exp.unwrapCause()))) {
```
can we have an explicit boolean for this? feels hacky...

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -566,13 +528,70 @@ void finishOnRemoteSuccess(Response response) {
+        PrimaryPhase(Request request, TransportChannel channel) {
+            this.observer = new ClusterStateObserver(clusterService, request.timeout(), logger);
+            final String concreteIndex;
+            if (resolveIndex()) {
```
I wonder if we should resolve things again here. I think we should make it as simple as possible here. Maybe have a Primary request (or use internal request) which has the shardid already built into it. All we need to do in the action is lookup the index shard, see that we have it and do our thing.

areek commented:
@bleskes I have updated the PR to:
  - allow `TransportReplicationAction`subclasses to only provide `ShardID` instead of `ShardIterator`
  - simplified `PrimaryPhase` to work with `ShardID`

Though this is still a WIP (have to cleanup the `ReplicationPhase` logic around replicating to only non-local copies), would appreciate a review.

bleskes commented:
@areek I think this is going in the right direction. left comments here and there on the commit.

areek commented:
Thanks @bleskes for the feedback. I updated the PR to address them. 
Now subclasses for `TransportReplicationAction` sets the target shardID in `resolveRequest` and the base class is responsible for routing to the primary copy and replicating to all the replica copies according to the cluster's routing table instead of relying on the subclasses to provide shard routings for replica copies.

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/flush/TransportShardFlushAction.java:
```diff
@@ -81,18 +79,18 @@ protected boolean checkWriteConsistency() {
-    protected ClusterBlockException checkRequestBlock(ClusterState state, InternalRequest request) {
-        return state.blocks().indicesBlockedException(ClusterBlockLevel.METADATA_WRITE, new String[]{request.concreteIndex()});
+    protected void resolveRequest(ClusterState state, String concreteIndex, ShardFlushRequest request) {
+        // request shardID already resolved at request construction
```
does it make sense to assert on it?

areek replied:
There is an assert for this in the base class, wouldn't it be redundant?

bleskes replied:
yeah, saw it later. Agreed it's OK.

bleskes commented on core/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:
```diff
@@ -275,7 +275,7 @@ private void executeBulk(final BulkRequest bulkRequest, final long startTime, fi
                     }
                 } else {
-                    ShardId shardId = clusterService.operationRouting().deleteShards(clusterState, concreteIndex, deleteRequest.type(), deleteRequest.id(), deleteRequest.routing()).shardId();
+                    ShardId shardId = clusterService.operationRouting().indexShards(clusterState, concreteIndex, deleteRequest.id(), deleteRequest.routing()).shardId();
```
good these variants go away...

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -125,29 +127,37 @@ protected TransportReplicationAction(Settings settings, String actionName, Trans
+    /**
+     * Cluster level block to check before request execution
+     */
+    protected ClusterBlockLevel globalBlockLevel() {
```
++

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -294,15 +309,34 @@ public void onTimeout(TimeValue timeout) {
+                    failReplicaIfNeeded(t);
                 } catch (Throwable unexpected) {
-                    logger.error("{} unexpected error while failing replica", request.internalShardId.id(), unexpected);
+                    logger.error("{} unexpected error while failing replica", request.shardId().id(), unexpected);
```
exception is in the wrong location :( (was already the case)

areek replied:
Good catch. fixed

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -314,23 +348,13 @@ protected void responseWithFailure(Throwable t) {
         }
     }
 
-    protected class PrimaryOperationRequest {
```
nice that this is removed.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -368,35 +392,50 @@ public void onFailure(Throwable e) {
         @Override
         protected void doRun() {
+            // shardID added to the request, as a side effect of calling resolveRequest
             if (checkBlocks() == false) {
```
can we change the method name to resolveRequestAndCheckBlocks? I had a worry that the request is not resolved by the time we check the index routing table based on request.shardId.

areek replied:
I renamed it to `checkBlocksAndResolveRequest` to clarify the ordering of the routines.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -368,35 +392,50 @@ public void onFailure(Throwable e) {
-                retryBecauseUnavailable(shardIt.shardId(), "No active shards.");
+            final IndexRoutingTable indexRoutingTable = observer.observedState().getRoutingTable().index(request.shardId().getIndex());
+            if (indexRoutingTable == null) {
+                logger.trace("index for shard [{}] is not found, scheduling a retry.", request.shardId());
```
can we add the action name and the request to all of these trace logs? see https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java#L586

areek replied:
Added trace logging in https://github.com/areek/elasticsearch/commit/9ca6209503fb30331fbb275130f99db5e9c3b40f

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -368,35 +392,50 @@ public void onFailure(Throwable e) {
+            final IndexRoutingTable indexRoutingTable = observer.observedState().getRoutingTable().index(request.shardId().getIndex());
+            if (indexRoutingTable == null) {
+                logger.trace("index for shard [{}] is not found, scheduling a retry.", request.shardId());
+                retryBecauseUnavailable(request.shardId(), "index is not active");
```
this doesn't mean the index is not active, but rather that it doesn't exist or is closed. I don't think we need to retry in that case. [Old cold would throw `IndexNotFoundException` in this case](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/cluster/routing/OperationRouting.java#L203).

areek replied:
Great catch! To restore the behaviour for unknown/closed index and unknown shard, now we use `OperationRouting#shards` instead of using the routing table directly. Fixed in https://github.com/areek/elasticsearch/commit/e22658dd07eae4439f293a182c8a6a09e3e23921

bleskes replied:
cool. That shards method is really a bad name (I know it was there). maybe rename to shardRoutingTable?

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -368,35 +392,50 @@ public void onFailure(Throwable e) {
+            final IndexShardRoutingTable shardRoutingTable = indexRoutingTable.shard(request.shardId().getId());
+            if (shardRoutingTable == null) {
+                logger.trace("routing for shard [{}] is not found, scheduling a retry.", request.shardId());
+                retryBecauseUnavailable(request.shardId(), "primary shard is not active");
```
same comment. We shouldn't retry in this case. [Old code throws ShardNotFoundException](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/cluster/routing/OperationRouting.java#L230)

areek replied:
Fixed in areek@e22658d

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -424,79 +457,50 @@ protected boolean checkBlocks() {
             }
+            // request does not have a shardId yet, we need to pass the concrete index to resolve shardId
+            resolveRequest(clusterState, concreteIndex, request);
+            assert request.shardId() != null : "request shardID must be set in resolveRequest";
```
++ on the assert

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -424,79 +457,50 @@ protected boolean checkBlocks() {
-                        }
+        private void performAction(final String nodeId, final String action) {
+            DiscoveryNode node = observer.observedState().nodes().get(nodeId);
+            final boolean isPrimaryAction = action.equals(transportPrimaryAction);
```
Nit: I think it will be safer to have this boolean as a parameter and determine the action here. I'm weary of arbitrary string input.

areek replied:
changed in https://github.com/areek/elasticsearch/commit/4ffdf5846f8b8941c5c23053dc902f9969ad836a

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -424,79 +457,50 @@ protected boolean checkBlocks() {
+                        // if we got disconnected from the node, or the node / shard is not in the right state (being closed)
+                        if (exp.unwrapCause() instanceof ConnectTransportException || exp.unwrapCause() instanceof NodeClosedException ||
+                                (isPrimaryAction && retryPrimaryException(exp.unwrapCause()))) {
+                            // we already marked it as started when we executed it (removed the listener) so pass false
```
left over unneeded comment.

areek replied:
removed in https://github.com/areek/elasticsearch/commit/4ffdf5846f8b8941c5c23053dc902f9969ad836a

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -566,42 +556,61 @@ void finishOnRemoteSuccess(Response response) {
-        void performOnPrimary(final ShardRouting primary, final ShardIterator shardsIt) {
-            final String writeConsistencyFailure = checkWriteConsistency(primary);
+        void retryBecauseUnavailable(ShardId shardId, String message) {
+            retry(new UnavailableShardsException(shardId, message + " Timeout: [" + request.timeout() + "], request: " + request.toString()));
```
we can now use the fancy exception message formatting.

areek replied:
fancied up exception message formatting in https://github.com/areek/elasticsearch/commit/9ca6209503fb30331fbb275130f99db5e9c3b40f

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -566,42 +556,61 @@ void finishOnRemoteSuccess(Response response) {
+                indexShardReference = getIndexShardOperationsCounter(shardId);
+                Tuple<Response, ReplicaRequest> primaryResponse = shardOperationOnPrimary(clusterState, request);
+                logger.trace("operation completed on primary [{}]", shardId);
+                // we cache meta data to resolve settings even if the index gets deleted after primary operation
```
this feels a bit weird. Can you elaborate on what happened?

areek replied:
This was to prevent a NPE on index metadata used to decide whether we wanted to execute actions on replica (for shadow replicas) and grab the indexUUID at the replication phase. This is possible when the index has been deleted after the primary operation has completed and before the replication phase has begun. 

I have cleaned this up and now the case is handled in ReplicationPhase ctor. Now if the index gets deleted before replication, the replication phase is skipped.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -566,42 +556,61 @@ void finishOnRemoteSuccess(Response response) {
                 if (ExceptionsHelper.status(e) == RestStatus.CONFLICT) {
                     if (logger.isTraceEnabled()) {
-                        logger.trace(primary.shortSummary() + ": Failed to execute [" + internalRequest.request() + "]", e);
+                        logger.trace("Failed to execute [" + request + "]", e);
```
why did we loose the primary.shortSummary?

areek replied:
This method uses the resolved shard id from the request rather than obtaining it from the shard routing as before. currently, we log the exception along with the request and the resolved shardID. Though we can get the shard routing from the state, IMO all relevant information is already logged. WDYT?

bleskes replied:
I see.Yeah, agreed that would be an overkill. I'm fine with leaving as is.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -566,42 +556,61 @@ void finishOnRemoteSuccess(Response response) {
                 } else {
                     if (logger.isDebugEnabled()) {
-                        logger.debug(primary.shortSummary() + ": Failed to execute [" + internalRequest.request() + "]", e);
+                        logger.debug("Failed to execute [" + request + "]", e);
```
same

areek replied:
see comment above

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -670,134 +706,61 @@ protected Releasable getIndexShardOperationsCounter(ShardId shardId) {
+            this.executeOnReplica = shouldExecuteReplication(indexMetaData.getSettings());
+            this.indexUUID = indexMetaData.getIndexUUID();
+
+            // we get a new state to route replication operations based on the latest shard routings and states
```
can we add to this: 

"we have to do this after succesfully indexing into the primary in order to honour recovery semantics. we have to make sure that every operation indexed into the primary after recovery start will also be replicated to the recovery target. If we use an old cluster state, we may miss a relocation that has started since then."

areek replied:
slightly modified and added in https://github.com/areek/elasticsearch/commit/e22658dd07eae4439f293a182c8a6a09e3e23921

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -670,134 +706,61 @@ protected Releasable getIndexShardOperationsCounter(ShardId shardId) {
+            // we calculate number of target nodes to send replication operations, including nodes with relocating shards
+            int numberOfIgnoredShardInstances = 0;
             int numberOfPendingShardInstances = 0;
-            if (observer.observedState() != newState) {
```
w00t.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -670,134 +706,61 @@ protected Releasable getIndexShardOperationsCounter(ShardId shardId) {
+
+            // we get a new state to route replication operations based on the latest shard routings and states
+            this.clusterState = clusterService.state();
+            this.shardIt = clusterService.operationRouting().shards(clusterState, shardId.getIndex(), shardId.id()).shardsIt();
```
I don't think operationRouting adds any value here (it's just for resolving id / routing to shard id) . Can we just access the routing table directly?

areek replied:
Right, changed to access the routing table directly instead in https://github.com/areek/elasticsearch/commit/e22658dd07eae4439f293a182c8a6a09e3e23921

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -670,134 +706,61 @@ protected Releasable getIndexShardOperationsCounter(ShardId shardId) {
-                        numberOfPendingShardInstances += 2;
-                    } else {
+            String localNodeId = clusterState.nodes().localNodeId();
+            for (ShardRouting shard : shardIt.asUnordered()) {
```
let's just skip the iterator and use IndexShardRoutingTable.shards() (put it in a variable and use it in do run as well)

areek replied:
done

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -840,94 +803,67 @@ protected void doRun() {
+
+                // we never execute replication operation locally as primary operation has already completed locally
+                // hence, we ignore any local shard for replication
+                if (clusterState.nodes().localNodeId().equals(shard.currentNodeId()) == false) {
```
can we cache the localNodeId?

areek replied:
I cached the state's discovery node instead. This minimizes the use of the cluster state in replication phase to only its ctor.

bleskes replied:
++

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -840,94 +803,67 @@ protected void doRun() {
-                    }
+                }
+                // send operation to relocating shard if not local
+                if (shard.relocating() &&  clusterState.nodes().localNodeId().equals(shard.relocatingNodeId()) == false) {
```
we do we have the extra protection for the local relocation? When is it needed?

areek replied:
the extra protection is not needed now, but makes replication action replicate to all shards except the shard where the primary action has been executed. In the future, we might want to execute a primary action on the relocating node instead of the current primary node and have the replication take care of operation on the current primary.

bleskes commented on core/src/main/java/org/elasticsearch/action/termvectors/TransportMultiTermVectorsAction.java:
```diff
@@ -80,7 +79,7 @@ protected void doExecute(final MultiTermVectorsRequest request, final ActionList
                 continue;
             }
             ShardId shardId = clusterService.operationRouting().getShards(clusterState, concreteSingleIndex,
```
get shards here is an overkill. Let's add a operationRouting().resolveShard() method that returns a shardId.

areek replied:
Agreed. I changed it to use OperationRouting#shardId instead.

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/OperationRouting.java:
```diff
@@ -54,16 +53,12 @@ public OperationRouting(Settings settings, AwarenessAllocationDecider awarenessA
 
-    public ShardIterator indexShards(ClusterState clusterState, String index, String type, String id, @Nullable String routing) {
-        return shards(clusterState, index, type, id, routing).shardsIt();
+    public ShardIterator indexShards(ClusterState clusterState, String index, String id, @Nullable String routing) {
```
I think we can reduce the scope of this change by exposing a resolveShardId method that resolves index,type,id and routing to a shardId. And then we don't need to touch these. Also, I know that type is not used now, but why not pass it? is there a place we don't have it? I hope we can back port this change to 2.x, so having type here will reduce the change.

bleskes replied:
OK. just saw the shardId() method. Good. the rest still holds ;)

bleskes replied:
ping about the rest?

bleskes commented on core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java:
```diff
@@ -172,29 +168,68 @@ public void testNotStartedPrimary() throws InterruptedException, ExecutionExcept
         assertIndexShardCounter(1);
     }
 
+    public void testPrimaryAction() throws InterruptedException, ExecutionException {
```
what are testing here? sounds like primaryPhaseExecutesRequest

bleskes commented on core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java:
```diff
@@ -172,29 +168,68 @@ public void testNotStartedPrimary() throws InterruptedException, ExecutionExcept
+        clusterService.setState(state(index, randomBoolean(), ShardRoutingState.STARTED, ShardRoutingState.STARTED));
+        Request request = new Request(shardId).timeout("1ms");
+        PlainActionFuture<Response> listener = new PlainActionFuture<>();
+        TransportReplicationAction.ReroutePhase primaryPhase = action.new ReroutePhase(request, listener);
```
primaryPhase->reroutePhase

bleskes commented on core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java:
```diff
@@ -172,29 +168,68 @@ public void testNotStartedPrimary() throws InterruptedException, ExecutionExcept
+        assertThat("request was not processed on primary", request.processedOnPrimary.get(), equalTo(true));
+    }
+
+    public void testRerouteResolve() throws Exception {
```
wondering - how is this test different from testRoutingToPrimary ?

bleskes commented on core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java:
```diff
@@ -172,29 +168,68 @@ public void testNotStartedPrimary() throws InterruptedException, ExecutionExcept
         final ShardId shardId = new ShardId(index, 0);
 
-        clusterService.setState(stateWithStartedPrimary(index, randomBoolean(), 3));
+        clusterService.setState(state(index, true, ShardRoutingState.STARTED, ShardRoutingState.STARTED));
```
why this simplification in cluster state structure?

bleskes commented on core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java:
```diff
@@ -232,7 +258,7 @@ public void testWriteConsistency() throws ExecutionException, InterruptedExcepti
         final int totalShards = 1 + assignedReplicas + unassignedReplicas;
         final boolean passesWriteConsistency;
-        Request request = new Request(shardId).consistencyLevel(randomFrom(WriteConsistencyLevel.values()));
+        Request request = new Request(shardId).consistencyLevel(randomFrom(WriteConsistencyLevel.values())).setShardId(shardId);
```
double set shardId?

bleskes commented on core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java:
```diff
@@ -422,7 +447,7 @@ public void testCounterOnPrimary() throws InterruptedException, ExecutionExcepti
                 ShardRoutingState.STARTED));
         logger.debug("--> using initial state:\n{}", clusterService.state().prettyPrint());
-        Request request = new Request(shardId).timeout("100ms");
+        Request request = new Request(shardId).timeout("100ms").setShardId(shardId);
```
same double set of shardId?

bleskes commented on plugins/lang-groovy/src/test/java/org/elasticsearch/messy/tests/IndicesRequestTests.java:
```diff
@@ -178,7 +178,7 @@ public void testAnalyze() {
 
     public void testIndex() {
-        String[] indexShardActions = new String[]{IndexAction.NAME, IndexAction.NAME + "[r]"};
+        String[] indexShardActions = new String[]{IndexAction.NAME + "[p]", IndexAction.NAME + "[r]"};
```
don't we want to keep on testing `IndexAction.NAME` here?

bleskes replied:
ping?

areek replied:
Good catch, added back testing for the reroute phase for index, delete, refresh and flush actions.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/ReplicationRequest.java:
```diff
@@ -61,6 +61,15 @@ public ReplicationRequest(ActionRequest request) {
+    public ReplicationRequest(ActionRequest request, ShardId shardId) {
+        super(request);
+        this.index = shardId.getIndex();
+        this.internalShardId = shardId;
```
can we rename internalShardId to resolvedShardId ? also add some java docs on how it should be used. I think that , as a separate change we can move the index field to another class DocReplicationRequest that is a subclass of this and a parent of IndexRequest and friends. It will be cleaner. No need to do it now.

areek replied:
I renamed internalShardId to resolvedShardId and added javadocs to clarify how it is used in https://github.com/areek/elasticsearch/commit/8e7d9b5894e569a47fa29a23d8d193e1aac3ffe9

bleskes commented:
I did a full review again and I really like where this is going. I would love it if @jasontedor gives a look as well.

areek commented:
I pushed the current changes to a feature branch `feature/refactor_replication_action` for CI. I will update the branch with any new commits here

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -670,135 +702,73 @@ protected Releasable getIndexShardOperationsCounter(ShardId shardId) {
+            final IndexRoutingTable index = state.getRoutingTable().index(shardId.getIndex());
+            final IndexShardRoutingTable shardRoutingTable = (index != null) ? index.shard(shardId.id()) : null;
+            final IndexMetaData indexMetaData = state.getMetaData().index(shardId.getIndex());
+            this.shards = (shardRoutingTable != null) ? shardRoutingTable.shards() : Collections.emptyList();
```
I think this is the right thing but can we log a debug level log here? this is extremely exceptional for people to index into and delete their index concurrently. We should have some visibility into it. As it is strictly speaking OK from an API perspective (people can do that), I wouldn't go with WARN/INFO, but with DEBUG

areek replied:
This is a good idea. I added debug logging in https://github.com/areek/elasticsearch/commit/688197f0be6050c4a63bd4cb71de88cc7dd29524 to add visibility when replication phase is entirely skipped due to index deletion after the primary operation.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -837,97 +807,72 @@ protected void doRun() {
-                    }
+                }
+                // send operation to relocating shard if not local
+                if (shard.relocating() &&  nodes.localNodeId().equals(shard.relocatingNodeId()) == false) {
```
responding to previous comments here, so we won't loose the discussion

>> Why do we have the extra protection for the local relocation? When is it needed?

> the extra protection is not needed now, but makes replication action replicate to all shards except the shard where the primary action has been executed. In the future, we might want to execute a primary action on the relocating node instead of the current primary node and have the replication take care of operation on the current primary.

I see, but in the case still want to replicate to the source shard. I suggest we do all of that in a followup change, instead of trying to anticipate it now. Makes sense?

areek replied:
Makes sense, I reverted this change and now only check if the current node of a replica copy is non-local for replication

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/OperationRouting.java:
```diff
@@ -219,12 +214,12 @@ protected IndexRoutingTable indexRoutingTable(ClusterState clusterState, String
     }
 
-    protected IndexShardRoutingTable shards(ClusterState clusterState, String index, int shardId) {
+    public IndexShardRoutingTable shards(ClusterState clusterState, String index, int shardId) {
```
can we call this shardRoutingTable?, if we expose it to the public, I wonder if we should just put `shardRoutingTable` method on the routing table , as a shortcut. Wouldn't it make more sense?

areek replied:
I added a shardRoutingTable shortcut to RoutingTable, which returns the shard routing table and throws appropriate exceptions when the index and/or shard id is unknown in https://github.com/areek/elasticsearch/commit/688197f0be6050c4a63bd4cb71de88cc7dd29524.

bleskes replied:
looks good.

areek commented:
Thanks @bleskes for the feedback :) I updated the PR to address them all and added some unit tests for the reroute and shard operation phases in replication action. A review would be awesome.

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/flush/TransportShardFlushAction.java:
```diff
@@ -61,16 +58,16 @@ protected ActionWriteResponse newResponseInstance() {
-    protected Tuple<ActionWriteResponse, ShardFlushRequest> shardOperationOnPrimary(ClusterState clusterState, PrimaryOperationRequest shardRequest) throws Throwable {
-        IndexShard indexShard = indicesService.indexServiceSafe(shardRequest.shardId.getIndex()).getShard(shardRequest.shardId.id());
-        indexShard.flush(shardRequest.request.getRequest());
+    protected Tuple<ActionWriteResponse, ShardFlushRequest> shardOperationOnPrimary(MetaData metaData, ShardFlushRequest shardRequest) throws Throwable {
```
IMO this is over doing this. I think passing just the meta data is over restrictive. ClusterState feels more natural. I would prefer this stays as ClusterState

areek replied:
The only way implementations use the passed in clusterstate to shard operation is to figure out if explicit routing is required (in index and bulk requests), resolve the routing and to know the index creation version from the metadata. IMO, passing in the entire cluster state is redundant whereas passing in the metadata hints at using it to validate the request before performing any operations. I might be missing something though. Do we really need to pass the entire cluster state here?

bleskes replied:
> whereas passing in the metadata hints at using it to validate the request before performing any operations

I'm not sure this hints what you suggest.

To me cluster state is simple  in terms of thinking - you get everything, do what ever you want. No a biggy though. If you feel strongly about this, let's just keep it.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/ReplicationRequest.java:
```diff
@@ -42,7 +42,12 @@
+     * shard id gets resolved by the transport action before performing request operation
+     * and at request creation time for shard-level bulk, refresh and flush requests.
+     */
+    protected ShardId resolvedShardId;
```
I'm sorry for flip-flopping on this, but thinking about this again (with the new model we introduced), I think it's simpler to call this just shardId , which must be set either at construction time (for things like refresh/bulk/flush actions) or when the request is resolved. Feels unnatural to call this resolvedShardId when it can be part of the constructor.

areek replied:
I like naming this `shardId` instead of `resolvedShardId`, changed in https://github.com/areek/elasticsearch/commit/fd2b09211f73e801b1e8f4bf51ba4f5cc19ec347

bleskes commented:
LGTM except some minor requests. Thanks areek for all the hard work. Let's see what @jasontedor has to say :)

jasontedor commented on core/src/main/java/org/elasticsearch/action/admin/indices/flush/ShardFlushRequest.java:
```diff
@@ -53,5 +54,8 @@ public void writeTo(StreamOutput out) throws IOException {
-
+    @Override
+    public String toString() {
+        return "flush {" + shardId() + "}";
```
I think this should also concatenate the underlying `FlushRequest` `request#toString`.

areek replied:
I added a simple toString method to ReplicationRequest in https://github.com/areek/elasticsearch/commit/6fa8140adf3b9c774431b5f815543930d144cde2 for providing the index and shard info if resolved or just the index name if the request still needs resolving. Note that I intentionally didn't use this for index and delete requests.

jasontedor commented on core/src/main/java/org/elasticsearch/action/admin/indices/flush/TransportShardFlushAction.java:
```diff
@@ -81,18 +78,18 @@ protected boolean checkWriteConsistency() {
-    protected ClusterBlockException checkRequestBlock(ClusterState state, InternalRequest request) {
-        return state.blocks().indicesBlockedException(ClusterBlockLevel.METADATA_WRITE, new String[]{request.concreteIndex()});
+    protected void resolveRequest(MetaData metaData, String concreteIndex, ShardFlushRequest request) {
+        // request shardID already resolved at request construction
```
I wonder if this should just be the implementation provided in `TransportReplicationAction`? It appears there are only two classes that currently provide a non-trivial implementation of this method.

areek replied:
That is a good idea. I made `resolveRequest` a no-op by default in https://github.com/areek/elasticsearch/commit/29397c1670fc935dda4e06a4b33ab454f9fe3f80

jasontedor commented on core/src/main/java/org/elasticsearch/action/bulk/BulkShardRequest.java:
```diff
@@ -93,4 +91,9 @@ public void readFrom(StreamInput in) throws IOException {
+
+    @Override
+    public String toString() {
+        return "shard bulk {" + shardId() + "}";
```
Same thing. As in `ShardFlushRequest`; I think the underlying request should be concatenated here.

areek replied:
fixed in https://github.com/areek/elasticsearch/commit/6fa8140adf3b9c774431b5f815543930d144cde2

jasontedor commented on core/src/main/java/org/elasticsearch/cluster/routing/OperationRouting.java:
```diff
@@ -219,22 +207,20 @@ protected IndexRoutingTable indexRoutingTable(ClusterState clusterState, String
 
     // either routing is set, or type/id are set
```
I think this comment is stale now?

jasontedor commented on core/src/main/java/org/elasticsearch/cluster/routing/OperationRouting.java:
```diff
@@ -55,19 +54,16 @@ public OperationRouting(Settings settings, AwarenessAllocationDecider awarenessA
     }
 
     public ShardIterator indexShards(ClusterState clusterState, String index, String type, String id, @Nullable String routing) {
```
The type parameter is unused now. Can it be removed?

bleskes replied:
it's used in 2.x and asked Areek to keep it to limit the scope of the change as I hope we can back port it.

bleskes replied:
@areek can we remove all the changes to this class now? or is there some stuff here we need?

jasontedor replied:
> it's used in 2.x and asked Areek to keep it to limit the scope of the change as I hope we can back port it.

ACK.

areek replied:
@bleskes we introduced the public `shardId` method here, so we can't remove all the changes to this class. I looked into 2.x and I think we can back port it without any issues.

jasontedor commented on core/src/main/java/org/elasticsearch/cluster/routing/OperationRouting.java:
```diff
@@ -55,19 +54,16 @@ public OperationRouting(Settings settings, AwarenessAllocationDecider awarenessA
+        return shards(clusterState, index, id, routing).shardsIt();
     }
 
     public ShardIterator getShards(ClusterState clusterState, String index, String type, String id, @Nullable String routing, @Nullable String preference) {
```
This `type` parameter is also unused now. Can it be removed?

areek replied:
it can be but not removing to have a easier time backporting this to 2.x

jasontedor replied:
> it can be but not removing to have a easier time backporting this to 2.x

ACK.

jasontedor commented on core/src/main/java/org/elasticsearch/action/support/replication/ReplicationRequest.java:
```diff
@@ -164,15 +178,18 @@ public void readFrom(StreamInput in) throws IOException {
     public void writeTo(StreamOutput out) throws IOException {
         super.writeTo(out);
-        out.writeOptionalStreamable(internalShardId);
+        out.writeOptionalStreamable(shardId);
```
Nit: I prefer when the `readFrom`/`writeTo` are symmetrical. Specifically, since we can't use `readOptionalStreamable` in `readFrom` due to `ShardId` only having a private constructor, I think that we should have a symmetrical `if` here in the `writeTo`.

areek replied:
I fixed it in the feature branch pulling in the changes to readFrom/writeTo methods already.  https://github.com/areek/elasticsearch/commit/d6df74baebd52d26e0200d818e0e65fec442a749 fixes this for this PR.

jasontedor commented on core/src/main/java/org/elasticsearch/action/support/replication/ReplicationRequest.java:
```diff
@@ -154,7 +168,7 @@ public ActionRequestValidationException validate() {
         if (in.readBoolean()) {
-            internalShardId = ShardId.readShardId(in);
+            shardId = ShardId.readShardId(in);
         }
```
There should be an `else` clause to set `shardId` to `null` if the stream indicates there is no `ShardId` in the stream.

areek replied:
fixed in https://github.com/areek/elasticsearch/commit/d6df74baebd52d26e0200d818e0e65fec442a749

jasontedor commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -368,35 +407,44 @@ public void onFailure(Throwable e) {
+        protected boolean checkBlocksAndResolveRequest() {
+            final ClusterState clusterState = observer.observedState();
+            ClusterBlockException blockException = clusterState.blocks().globalBlockedException(globalBlockLevel());
             if (blockException != null) {
```
The handling here is duplicated in the handling after checking the index blocks. Can we collapse this logic?

areek replied:
Right. I added a method to re-use handling block exceptions in https://github.com/areek/elasticsearch/commit/df36946a182961799f73f353436846cba1d97057

jasontedor commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -424,79 +466,47 @@ protected boolean checkBlocks() {
-        }
+        private void performAction(final String nodeId, final String action, final boolean isPrimaryAction) {
+            DiscoveryNode node = observer.observedState().nodes().get(nodeId);
+            transportService.sendRequest(node, action, request, transportOptions, new BaseTransportResponseHandler<Response>() {
```
Check that `node` is not `null`?

areek replied:
I don't see how getting a `null` is possible here. We already check whether the node exists for the state in `ReroutePhase#doRun`. Am I missing something?

jasontedor replied:
I saw the check, and maybe I'm making a mistake but can't the observed cluster state have changed between then and now?

areek replied:
The observed state can only change when we retry (via `ClusterStateObserver#waitForNextChange`) and on retry we run `ReroutePhase#doRun` which checks if the node exists for the new state. Hence I don't think the observed cluster state can change from after checking if the node exists to performing an action on the node.

jasontedor replied:
Okay, you're correct. Can we make this clearer? Grab a copy of the observed cluster state at the start of doRun, pass it along to the global checks and request resolving, and finally on its merry way to the rest of the request handling? On a retry, that whole chain will start over after a new observed state is captured?

areek replied:
I like this a lot! I restructured the doRun method in https://github.com/areek/elasticsearch/commit/1159d3ead370dcc94835b5b5146f7344eef65a0a to clarify that we use one copy of the state throughout the reroute phase. This change also nuked the  `checkBlocksAndResolveRequest` with side-effect that was pointed out in another comment.

bleskes replied:
side note: the whole point of the observer was to sample the cluster at known moments and be consistent. I agree the new code is simpler so change is good. @jasontedor if you have a better suggestion for doc/naming in the cluster state observer please let me know.

jasontedor commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -368,35 +407,44 @@ public void onFailure(Throwable e) {
         protected void doRun() {
-            if (checkBlocks() == false) {
+            // shardID added to the request, as a side effect of calling resolveRequest
+            if (checkBlocksAndResolveRequest() == false) {
```
Is there a good reason for combining these checks? I'm not a fan of the side-effectful request mutating (but I understand why it's done), but at least can it be brought into `doRun`?

areek replied:
I agree it would be clearer to perform the block checks and resolve request in the `doRun` method itself, https://github.com/areek/elasticsearch/commit/1159d3ead370dcc94835b5b5146f7344eef65a0a fixes this and makes it clear that we only operate on one copy of the state throughout a single run of the reroute phase.

jasontedor replied:
This is wonderful! Now `observer#observedState()` is invoked in exactly one place in the source file.

jasontedor commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -548,60 +545,79 @@ void finishAsFailed(Throwable failure) {
+                }
                 listener.onResponse(response);
             } else {
                 assert false : "finishOnRemoteSuccess called but operation is already finished";
```
I think that `finishOnRemoteSuccess` should be changed to `finishOnSuccess` too?

areek replied:
Good catch, fixed in https://github.com/areek/elasticsearch/commit/df36946a182961799f73f353436846cba1d97057

jasontedor commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -314,23 +363,17 @@ protected void responseWithFailure(Throwable t) {
-            try (Releasable shardReference = getIndexShardOperationsCounter(request.internalShardId)) {
-                shardOperationOnReplica(request.internalShardId, request);
+            assert request.shardId() != null : "request shardId must be set";
+            try (Releasable shardReference = getIndexShardOperationsCounter(request.shardId())) {
```
Nit: Can we rename `shardReference` to `ignored`? I don't know about Eclipse, but this will at least prevent IntelliJ from issuing unused variable warnings.

areek replied:
fixed in https://github.com/areek/elasticsearch/commit/f2d6a931a9fdd82b85c543f9d68be1e1f70399a0

jasontedor commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -368,135 +411,91 @@ public void onFailure(Throwable e) {
-                } else {
-                    finishAsFailed(blockException);
+                if (logger.isTraceEnabled()) {
+                    logger.trace("send action [{}] on primary [{}] for request [{}] with cluster state version [{} to [{}]]", actionName, request.shardId(), request, state.version(), primary.currentNodeId());
```
The logging brackets are off here: `[{} to [{}]]`.

areek replied:
good catch, fixed in https://github.com/areek/elasticsearch/commit/e54b84b8d59d249acc02cfa919cd5b8ee8b4677b

jasontedor commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -548,60 +534,79 @@ void finishAsFailed(Throwable failure) {
+        private final TransportChannel channel;
+        private final ClusterState state;
+        private final AtomicBoolean finished = new AtomicBoolean(false);
+        private volatile Releasable indexShardReference;
```
Why is this `volatile`? It doesn't look necessary to me.

areek replied:
It is not necessary, was left over fixed in https://github.com/areek/elasticsearch/commit/2c609d7f9d088847bde26fdf32ef395306d9e8d7

jasontedor commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -548,60 +534,79 @@ void finishAsFailed(Throwable failure) {
+        private final Request request;
+        private final TransportChannel channel;
+        private final ClusterState state;
+        private final AtomicBoolean finished = new AtomicBoolean(false);
```
Nit: The default constructor `AtomicBoolean#<init>()` can be used.

areek replied:
fixed in https://github.com/areek/elasticsearch/commit/2c609d7f9d088847bde26fdf32ef395306d9e8d7

jasontedor commented:
This is great; I really like this. The one hangup I continue to have is mutating the requests, but some of the concern there has been addressed by bringing it into the open in `ReroutePhase#doRun`.

LGTM.

areek commented:
Thanks for the thorough review, @bleskes and @jasontedor. I squashed all the commits and plan to merge this to master tomorrow.

areek commented:
back ported to 2.x in https://github.com/elastic/elasticsearch/commit/f1622aa60af3440591e0afddd892be30c07d96df
